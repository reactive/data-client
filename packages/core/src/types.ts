import { NormalizedIndex } from '@data-client/normalizr';
import type {
  UpdateFunction,
  AbstractInstanceType,
} from '@data-client/normalizr';
import type { ErrorTypes } from '@data-client/normalizr';

import type {
  ActionTypes,
  SetResponseAction,
  OptimisticAction,
} from './actions.js';
import type { Dispatch, Middleware, MiddlewareAPI } from './middlewareTypes.js';

export * from './actions.js';

export type { AbstractInstanceType, UpdateFunction };

export type PK = string;

/** Normalized state for Reactive Data Client
 *
 * @see https://dataclient.io/docs/concepts/normalization
 */
export interface State<T> {
  readonly entities: {
    readonly [entityKey: string]: { readonly [pk: string]: T } | undefined;
  };
  readonly endpoints: {
    readonly [key: string]: unknown | PK[] | PK | undefined;
  };
  readonly indexes: NormalizedIndex;
  readonly meta: {
    readonly [key: string]: {
      readonly date: number;
      readonly fetchedAt: number;
      readonly expiresAt: number;
      readonly prevExpiresAt?: number;
      readonly error?: ErrorTypes;
      readonly invalidated?: boolean;
      readonly errorPolicy?: 'hard' | 'soft' | undefined;
    };
  };
  readonly entitiesMeta: {
    readonly [entityKey: string]: {
      readonly [pk: string]: {
        readonly fetchedAt: number;
        readonly date: number;
        readonly expiresAt: number;
      };
    };
  };
  readonly optimistic: (SetResponseAction | OptimisticAction)[];
  readonly lastReset: number;
}

/** Singletons that handle global side-effects
 *
 * Kind of like useEffect() for the central data store
 *
 * @see https://dataclient.io/docs/api/Manager
 */
export interface Manager<Actions = ActionTypes> {
  /** @see https://dataclient.io/docs/api/Manager#getmiddleware */
  getMiddleware?(): Middleware<Actions>;
  /** @see https://dataclient.io/docs/api/Manager#middleware */
  middleware?: Middleware<Actions>;
  /** @see https://dataclient.io/docs/api/Manager#cleanup */
  cleanup(): void;
  /** @see https://dataclient.io/docs/api/Manager#init */
  init?: (state: State<any>) => void;
}

export type { Dispatch, Middleware, MiddlewareAPI };
