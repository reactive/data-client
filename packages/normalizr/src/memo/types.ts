import WeakDependencyMap from './WeakDependencyMap.js';
import { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';

export interface EntityCache
  extends Map<
    string,
    Map<
      string,
      WeakMap<EntityInterface, WeakDependencyMap<EntityPath, object, any>>
    >
  > {}

export type EndpointsCache = WeakDependencyMap<EntityPath, object, any>;

export type IndexPath = [key: string, field: string, value: string];
export type EntitySchemaPath = [key: string] | [key: string, pk: string];
export type QueryPath = IndexPath | EntitySchemaPath;
