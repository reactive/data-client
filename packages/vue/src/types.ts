import type { MaybeRefOrGetter } from 'vue';

/** Maps each parameter to accept raw value, Ref, ComputedRef, or getter */
export type MaybeRefsOrGetters<T extends readonly any[]> = {
  readonly [K in keyof T]: MaybeRefOrGetter<T[K]>;
};

/** Maps each parameter to accept raw value, Ref, ComputedRef, or getter, with nullable support */
export type MaybeRefsOrGettersNullable<T extends readonly any[]> = {
  readonly [K in keyof T]: MaybeRefOrGetter<T[K] | null>;
};
