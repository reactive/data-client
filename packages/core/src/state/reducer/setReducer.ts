import { normalize } from '@data-client/normalizr';

import Controller from '../../controller/Controller.js';
import type { State, SetAction } from '../../types.js';

export function setReducer(
  state: State<unknown>,
  action: SetAction,
  controller: Controller,
) {
  let value: any;
  if (typeof action.value === 'function') {
    const previousValue = controller.get(action.schema, ...action.args, state);
    if (previousValue === undefined) return state;
    value = action.value(previousValue);
  } else {
    value = action.value;
  }
  try {
    const { entities, indexes, entitiesMeta } = normalize(
      action.schema,
      value,
      action.args,
      state,
      action.meta,
    );
    return {
      entities,
      endpoints: state.endpoints,
      indexes,
      meta: state.meta,
      entitiesMeta,
      optimistic: state.optimistic,
      lastReset: state.lastReset,
    };
    // reducer must update the state, so in case of processing errors we simply compute the endpoints inline
  } catch (error: any) {
    // this is not always bubbled up, so let's double sure this doesn't fail silently
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      console.error(error);
    }
    return state;
  }
}
