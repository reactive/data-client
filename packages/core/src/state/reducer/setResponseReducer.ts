import { normalize } from '@data-client/normalizr';

import { OPTIMISTIC } from '../../actionTypes.js';
import AbortOptimistic from '../../controller/AbortOptimistic.js';
import type Controller from '../../controller/Controller.js';
import type {
  State,
  SetResponseAction,
  OptimisticAction,
} from '../../types.js';

export function setResponseReducer(
  state: State<unknown>,
  action: OptimisticAction | SetResponseAction,
  controller: Controller,
) {
  if (action.error) {
    return reduceError(state, action, action.response);
  }
  try {
    let response: any;
    // for true set's response is contained in action
    if (action.type === OPTIMISTIC) {
      // this should never happen
      /* istanbul ignore if */
      if (!action.endpoint.getOptimisticResponse) return state;
      try {
        // compute optimistic response based on current state
        response = action.endpoint.getOptimisticResponse.call(
          action.endpoint,
          controller.snapshot(state, action.meta.fetchedAt),
          ...action.args,
        );
      } catch (e: any) {
        // AbortOptimistic means 'do nothing', otherwise we count the exception as endpoint failure
        if (e.constructor === AbortOptimistic) {
          return state;
        }
        throw e;
      }
    } else {
      response = action.response;
    }
    const { result, entities, indexes, entitiesMeta } = normalize(
      action.endpoint.schema,
      response,
      action.args,
      state,
      action.meta,
    );
    const endpoints: Record<string, unknown> = {
      ...state.endpoints,
      [action.key]: result,
    };
    try {
      if (action.endpoint.update) {
        const updaters = action.endpoint.update(result, ...action.args);
        Object.keys(updaters).forEach(key => {
          endpoints[key] = updaters[key](endpoints[key]);
        });
      }
      // no reason to completely fail because of user-code error
      // integrity of this state update is still guaranteed
    } catch (error) {
      console.error(
        `The following error occured during Endpoint.update() for ${action.key}`,
      );
      console.error(error);
    }
    return {
      entities,
      endpoints,
      indexes,
      meta: {
        ...state.meta,
        [action.key]: {
          date: action.meta.date,
          fetchedAt: action.meta.fetchedAt,
          expiresAt: action.meta.expiresAt,
          prevExpiresAt: state.meta[action.key]?.expiresAt,
        },
      },
      entitiesMeta,
      optimistic: filterOptimistic(state, action),
      lastReset: state.lastReset,
    };
    // reducer must update the state, so in case of processing errors we simply compute the endpoints inline
  } catch (error: any) {
    if (typeof error === 'object') {
      error.message = `Error processing ${
        action.key
      }\n\nFull Schema: ${JSON.stringify(
        action.endpoint.schema,
        undefined,
        2,
      )}\n\nError:\n${error.message}`;
      if ('response' in action) error.response = action.response;
      error.status = 400;
    }

    // this is not always bubbled up, so let's double sure this doesn't fail silently
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      console.error(error);
    }
    return reduceError(state, action, error);
  }
}

function reduceError(
  state: State<unknown>,
  action: SetResponseAction | OptimisticAction,
  error: any,
): State<unknown> {
  if (error.name === 'AbortError') {
    // In case we abort simply undo the optimistic update and act like no fetch even occured
    // We still want those watching promises from fetch directly to observed the abort, but we don't want to
    // Trigger errors in this case. This means theoretically improperly built abortes useResource() could suspend forever.
    return {
      ...state,
      optimistic: filterOptimistic(state, action),
    };
  }
  return {
    ...state,
    meta: {
      ...state.meta,
      [action.key]: {
        date: action.meta.date,
        fetchedAt: action.meta.fetchedAt,
        expiresAt: action.meta.expiresAt,
        error,
        errorPolicy: action.endpoint.errorPolicy?.(error),
      },
    },
    optimistic: filterOptimistic(state, action),
  };
}
/** Filter all requests with same serialization that did not start after the resolving request */
function filterOptimistic(
  state: State<unknown>,
  resolvingAction: SetResponseAction | OptimisticAction,
) {
  return state.optimistic.filter(
    optimisticAction =>
      optimisticAction.key !== resolvingAction.key ||
      (optimisticAction.type === OPTIMISTIC ?
        optimisticAction.meta.fetchedAt !== resolvingAction.meta.fetchedAt
      : optimisticAction.meta.date > resolvingAction.meta.date),
  );
}
