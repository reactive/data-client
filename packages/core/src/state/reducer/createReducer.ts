import { expireReducer } from './expireReducer.js';
import { fetchReducer } from './fetchReducer.js';
import { invalidateReducer } from './invalidateReducer.js';
import { setReducer } from './setReducer.js';
import { setResponseReducer } from './setResponseReducer.js';
import {
  SET,
  INVALIDATE,
  RESET,
  FETCH,
  GC,
  OPTIMISTIC,
  INVALIDATEALL,
  EXPIREALL,
  SET_RESPONSE,
} from '../../actionTypes.js';
import type Controller from '../../controller/Controller.js';
import type { ActionTypes, State } from '../../types.js';

export default function createReducer(controller: Controller): ReducerType {
  return function reducer(
    state: State<unknown> | undefined,
    action: ActionTypes,
  ): State<unknown> {
    if (!state) state = initialState;
    switch (action.type) {
      case GC:
        // inline deletes are fine as these should have 0 refcounts
        action.entities.forEach(({ key, pk }) => {
          delete (state as any).entities[key]?.[pk];
          delete (state as any).entitiesMeta[key]?.[pk];
        });
        action.endpoints.forEach(fetchKey => {
          delete (state as any).endpoints[fetchKey];
          delete (state as any).meta[fetchKey];
        });
        return state;
      case FETCH:
        return fetchReducer(state, action);

      case OPTIMISTIC:
      // eslint-disable-next-line no-fallthrough
      case SET_RESPONSE:
        return setResponseReducer(state, action, controller);

      case SET:
        return setReducer(state, action, controller);

      case INVALIDATEALL:
      case INVALIDATE:
        return invalidateReducer(state, action);

      case EXPIREALL:
        return expireReducer(state, action);

      case RESET:
        return { ...initialState, lastReset: action.date };

      default:
        // A reducer must always return a valid state.
        // Alternatively you can throw an error if an invalid action is dispatched.
        return state;
    }
  } as any;
}

export const initialState: State<unknown> = {
  entities: {},
  endpoints: {},
  indexes: {},
  meta: {},
  entitiesMeta: {},
  optimistic: [],
  lastReset: 0,
};

type ReducerType = (
  state: State<unknown> | undefined,
  action: ActionTypes,
) => State<unknown>;

type Writable<T> = { [P in keyof T]: NonNullable<Writable<T[P]>> };
