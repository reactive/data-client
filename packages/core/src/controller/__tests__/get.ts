import { Entity, schema } from '@data-client/endpoint';

import { initialState } from '../../state/reducer/createReducer';
import { State } from '../../types';
import Controller from '../Controller';

class Tacos extends Entity {
  type = '';
  id = '';
}
const TacoList = new schema.Collection([Tacos]);
const entities = {
  Tacos: {
    1: { id: '1', type: 'foo' },
    2: { id: '2', type: 'bar' },
  },
  [TacoList.key]: {
    [TacoList.pk(undefined, undefined, '', [{ type: 'foo' }])]: ['1'],
    [TacoList.pk(undefined, undefined, '', [{ type: 'bar' }])]: ['2'],
    [TacoList.pk(undefined, undefined, '', [])]: ['1', '2'],
  },
};

describe('Controller.get()', () => {
  it('query Entity based on pk', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const taco = controller.get(Tacos, { id: '1' }, state);
    expect(taco).toBeDefined();
    expect(taco).toBeInstanceOf(Tacos);
    expect(taco).toMatchSnapshot();
    const taco2 = controller.get(Tacos, { id: '2' }, state);
    expect(taco2).toBeDefined();
    expect(taco2).toBeInstanceOf(Tacos);
    expect(taco2).not.toEqual(taco);
    // should maintain referential equality
    expect(taco).toBe(controller.get(Tacos, { id: '1' }, state));

    // @ts-expect-error
    () => controller.get(Tacos, { id: { bob: 5 } }, state);
    // @ts-expect-error
    expect(controller.get(Tacos, 5, state)).toBeUndefined();
    // @ts-expect-error
    () => controller.get(Tacos, { doesnotexist: 5 }, state);
  });

  describe('indexes', () => {
    class User extends Entity {
      id = '';
      username = '';
      staff = false;

      static indexes = ['username'] as const;
    }
    it('query Entity based on index', () => {
      const controller = new Controller();
      const state: State<unknown> = {
        ...initialState,
        entities: {
          User: {
            '1': { id: '1', username: 'bob' },
            '2': { id: '2', username: 'george' },
          },
        },
        indexes: {
          User: {
            username: {
              bob: '1',
              george: '2',
            },
          },
        },
      };

      const bob = controller.get(User, { username: 'bob' }, state);
      expect(bob).toBeDefined();
      expect(bob).toBeInstanceOf(User);
      expect(bob).toMatchSnapshot();
      // stability
      expect(controller.get(User, { username: 'bob' }, state)).toBe(bob);
      // should be same as id lookup
      expect(controller.get(User, { id: '1' }, state)).toBe(bob);
      // update index
      let nextState: State<unknown> = {
        ...state,
        entities: {
          ...state.entities,
          User: {
            ...state.entities.User,
            '1': { id: '1', username: 'george' },
            '2': { id: '2', username: 'bob' },
          },
        },
        indexes: {
          ...state.indexes,
          User: {
            ...state.indexes.User,
            username: {
              ...state.indexes.User.username,
              bob: '2',
              george: '1',
            },
          },
        },
      };
      expect(controller.get(User, { username: 'bob' }, nextState)).not.toBe(
        bob,
      );
      nextState = {
        ...state,
        entities: {
          ...state.entities,
          User: {
            ...state.entities.User,
            '1': { id: '1', username: 'bob', staff: true },
          },
        },
      };
      // update entity keep index
      const nextBob = controller.get(User, { username: 'bob' }, nextState);
      expect(nextBob).not.toBe(bob);
      expect(nextBob).toBeDefined();
      expect(nextBob).toBeInstanceOf(User);
      expect(nextBob?.staff).toBe(true);
    });

    it('query indexes after empty state', () => {
      const controller = new Controller();
      expect(
        controller.get(User, { username: 'bob' }, initialState),
      ).toBeUndefined();
      const state: State<unknown> = {
        ...initialState,
        entities: {
          User: {
            '1': { id: '1', username: 'bob' },
            '2': { id: '2', username: 'george' },
          },
        },
        indexes: {
          User: {
            username: {
              bob: '1',
              george: '2',
            },
          },
        },
      };
      const bob = controller.get(User, { username: 'bob' }, state);
      expect(bob).toBeDefined();
      expect(bob).toBeInstanceOf(User);
      expect(bob).toMatchSnapshot();
    });
  });

  it('query Collection based on args', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const tacos = controller.get(TacoList, { type: 'foo' }, state);
    expect(tacos).toBeDefined();
    expect(tacos?.[0]).toBeInstanceOf(Tacos);
    expect(tacos).toMatchSnapshot();
    const tacosBars = controller.get(TacoList, { type: 'bar' }, state);
    expect(tacosBars).toBeDefined();
    expect(tacosBars?.[0]).toBeInstanceOf(Tacos);
    expect(tacosBars).not.toEqual(tacos);
    // should maintain referential equality
    expect(tacos).toBe(controller.get(TacoList, { type: 'foo' }, state));

    const allTacos = controller.get(TacoList, state);
    expect(allTacos).toBeDefined();
    expect(allTacos).toHaveLength(2);
    expect(allTacos).toMatchSnapshot();

    // @ts-expect-error
    () => controller.get(TacoList, 5, state);
  });

  describe('query All', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const AllTacos = new schema.All(Tacos);

    it('should get all entities', () => {
      const allTacos = controller.get(AllTacos, state);
      expect(allTacos).toBeDefined();
      expect(allTacos).toHaveLength(2);
      expect(allTacos).toMatchSnapshot();

      // TODO: @ts-expect-error (we have a hack to make this not break other things now)
      () => controller.get(AllTacos, 5, state);
    });

    it('should maintain referential equality', () => {
      const allTacos = controller.get(AllTacos, state);
      expect(allTacos).toStrictEqual(controller.get(TacoList, state));
    });

    it('should include new entities when added', () => {
      controller.get(AllTacos, state);
      const tacosWithExtra = controller.get(AllTacos, {
        ...state,
        entities: {
          ...entities,
          Tacos: {
            ...entities.Tacos,
            3: { id: '3', type: 'extra' },
          },
        },
      });
      expect(tacosWithExtra).toBeDefined();
      expect(tacosWithExtra).toHaveLength(3);
      expect(tacosWithExtra).toMatchSnapshot();
    });
  });

  it('Query+All based on args', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const queryTacos = new schema.Query(
      new schema.All(Tacos),
      (tacos, { type }: { type?: string } = {}) => {
        if (!type) return tacos;
        return tacos.filter(taco => taco.type === type);
      },
    );

    const tacos = controller.get(queryTacos, { type: 'foo' }, state);
    expect(tacos).toBeDefined();
    expect(tacos?.[0]).toBeInstanceOf(Tacos);
    expect(tacos).toMatchSnapshot();
    const tacosBars = controller.get(queryTacos, { type: 'bar' }, state);
    expect(tacosBars).toBeDefined();
    expect(tacosBars?.[0]).toBeInstanceOf(Tacos);
    expect(tacosBars).not.toEqual(tacos);
    // should maintain referential equality
    expect(tacos).toBe(controller.get(queryTacos, { type: 'foo' }, state));

    const allTacos = controller.get(queryTacos, state);
    expect(allTacos).toBeDefined();
    expect(allTacos).toHaveLength(2);
    expect(allTacos).toMatchSnapshot();

    // @ts-expect-error
    () => controller.get(queryTacos, 5, state);

    const tacosWithExtra = controller.get(
      queryTacos,
      { type: 'bar' },
      {
        ...state,
        entities: {
          ...entities,
          Tacos: {
            ...entities.Tacos,
            3: { id: '3', type: 'bar', name: 'thing3' },
          },
        },
      },
    );
    expect(tacosWithExtra).toBeDefined();
    expect(tacosWithExtra).toHaveLength(2);
    expect(tacosWithExtra).toMatchSnapshot();
  });

  it('Query+Collection based on args', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const tacoCount = new schema.Query(TacoList, tacos => {
      return tacos.length ?? 0;
    });

    expect(controller.get(tacoCount, { type: 'foo' }, state)).toBe(1);
    expect(controller.get(tacoCount, { type: 'bar' }, state)).toBe(1);
    expect(controller.get(tacoCount, state)).toBe(2);
  });

  it('Union based on args', () => {
    class IDEntity extends Entity {
      id: string = '';
    }
    class User extends IDEntity {
      type = 'users';
      username: string = '';
    }
    class Group extends IDEntity {
      type = 'groups';
      groupname: string = '';
      memberCount = 0;
    }
    const queryPerson = new schema.Union(
      {
        users: User,
        groups: Group,
      },
      'type',
    );
    const controller = new Controller();
    const state = {
      ...initialState,
      entities: {
        User: {
          '1': { id: '1', type: 'users', username: 'bob' },
        },
        Group: {
          '2': { id: '2', type: 'groups', groupname: 'fast', memberCount: 5 },
        },
      },
    };
    const user = controller.get(queryPerson, { id: '1', type: 'users' }, state);
    expect(user).toBeDefined();
    expect(user).toBeInstanceOf(User);
    expect(user).toMatchSnapshot();
    const group = controller.get(
      queryPerson,
      { id: '2', type: 'groups' },
      state,
    );
    expect(group).toBeDefined();
    expect(group).toBeInstanceOf(Group);
    expect(group).toMatchSnapshot();

    // should maintain referential equality
    expect(user).toBe(
      controller.get(queryPerson, { id: '1', type: 'users' }, state),
    );

    // @ts-expect-error
    () => controller.get(queryPerson, { id: { bob: 5 } }, state);
    // @ts-expect-error
    expect(controller.get(queryPerson, 5, state)).toBeUndefined();
    // @ts-expect-error
    () => controller.get(queryPerson, { doesnotexist: 5 }, state);
    // @ts-expect-error
    () => controller.get(queryPerson, { id: '1', doesnotexist: 5 }, state);
  });

  it('Union based on args with function schemaAttribute', () => {
    class IDEntity extends Entity {
      id: string = '';
    }
    class User extends IDEntity {
      type = 'user';
      username: string = '';
    }
    class Group extends IDEntity {
      type = 'group';
      groupname: string = '';
      memberCount = 0;
    }
    const queryPerson = new schema.Union(
      {
        users: User,
        groups: Group,
      },
      (value: { type: 'users' | 'groups' }) => value.type,
    );
    const controller = new Controller();
    const state = {
      ...initialState,
      entities: {
        User: {
          '1': { id: '1', type: 'users', username: 'bob' },
        },
        Group: {
          '2': { id: '2', type: 'groups', groupname: 'fast', memberCount: 5 },
        },
      },
    };
    const user = controller.get(queryPerson, { id: '1', type: 'users' }, state);
    expect(user).toBeDefined();
    expect(user).toBeInstanceOf(User);
    expect(user).toMatchSnapshot();
    const group = controller.get(
      queryPerson,
      { id: '2', type: 'groups' },
      state,
    );
    expect(group).toBeDefined();
    expect(group).toBeInstanceOf(Group);
    expect(group).toMatchSnapshot();

    // should maintain referential equality
    expect(user).toBe(
      controller.get(queryPerson, { id: '1', type: 'users' }, state),
    );

    // these are the 'fallback case' where it cannot determine type discriminator, so just enumerates
    () => controller.get(queryPerson, { id: '1' }, state);
    // @ts-expect-error
    () => controller.get(queryPerson, { id: '1', type: 'notrealtype' }, state);
    // @ts-expect-error
    () => controller.get(queryPerson, { id: { bob: 5 }, type: 'users' }, state);
    // @ts-expect-error
    expect(controller.get(queryPerson, 5, state)).toBeUndefined();
    // @ts-expect-error
    () => controller.get(queryPerson, { doesnotexist: 5 }, state);
    // @ts-expect-error
    () => controller.get(queryPerson, { id: '1', doesnotexist: 5 }, state);
  });
});

describe('Snapshot.getQueryMeta()', () => {
  it('query Entity based on pk', () => {
    const controller = new Controller();
    const state = {
      ...initialState,
      entities,
    };
    const snapshot = controller.snapshot(state);
    const taco = snapshot.getQueryMeta(Tacos, { id: '1' }).data;
    expect(taco).toBeDefined();
    expect(taco).toBeInstanceOf(Tacos);
    expect(taco).toMatchSnapshot();
    const taco2 = snapshot.getQueryMeta(Tacos, { id: '2' }).data;
    expect(taco2).toBeDefined();
    expect(taco2).toBeInstanceOf(Tacos);
    expect(taco2).not.toEqual(taco);
    // should maintain referential equality
    expect(taco).toBe(snapshot.getQueryMeta(Tacos, { id: '1' }).data);

    // @ts-expect-error
    () => snapshot.getQueryMeta(Tacos, { id: { bob: 5 } });
    // @ts-expect-error
    expect(snapshot.getQueryMeta(Tacos, 5).data).toBeUndefined();
    // @ts-expect-error
    () => snapshot.getQueryMeta(Tacos, { doesnotexist: 5 });
  });
});
