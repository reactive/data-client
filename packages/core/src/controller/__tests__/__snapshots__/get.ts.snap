// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller.get() Query+All based on args 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`Controller.get() Query+All based on args 2`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
]
`;

exports[`Controller.get() Query+All based on args 3`] = `
[
  Tacos {
    "id": "2",
    "type": "bar",
  },
  Tacos {
    "id": "3",
    "name": "thing3",
    "type": "bar",
  },
]
`;

exports[`Controller.get() Union based on args 1`] = `
User {
  "id": "1",
  "type": "users",
  "username": "bob",
}
`;

exports[`Controller.get() Union based on args 2`] = `
Group {
  "groupname": "fast",
  "id": "2",
  "memberCount": 5,
  "type": "groups",
}
`;

exports[`Controller.get() query All should get all entities 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
]
`;

exports[`Controller.get() query All should include new entities when added 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
  Tacos {
    "id": "3",
    "type": "extra",
  },
]
`;

exports[`Controller.get() query Collection based on args 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`Controller.get() query Collection based on args 2`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
]
`;

exports[`Controller.get() query Entity based on index 1`] = `
User {
  "id": "1",
  "username": "bob",
}
`;

exports[`Controller.get() query Entity based on pk 1`] = `
Tacos {
  "id": "1",
  "type": "foo",
}
`;

exports[`Snapshot.getQueryMeta() query Entity based on pk 1`] = `
Tacos {
  "id": "1",
  "type": "foo",
}
`;
